### PR TITLE
Enable c++11 in the cuda backend. Remove boost dependency

### PR DESCRIPTION
**Pull request by umar456**

### CODE REVIEW

**Comment by pavanky** [comment]
We should perhaps wait until https://github.com/arrayfire/arrayfire/pull/1687 is merged.

File: src/backend/cuda/CMakeLists.txt [hunk 1]
```diff
@@ -104,7 +103,7 @@ IF(UNIX)
         # Enabling c++11 with nvcc 7.5 + gcc 6.x doesn't seem to work
         # Only solution for now is to force use c++03 for gcc 6.x
         IF(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "6.0.0")
-          SET(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS} -Xcompiler -std=c++98")
+          message( FATAL_ERROR "NVCC does not support GCC version 6.0 or greater." )
```
Comment (shehzan10) [inline src/backend/cuda/CMakeLists.txt]: Wait, what?
@pavanky I'm sure this concerns you.

Comment (pavanky) [inline src/backend/cuda/CMakeLists.txt]: @shehzan10 archlinux now packages gcc-5. Not an issue anymore.

Comment (shehzan10) [inline src/backend/cuda/CMakeLists.txt]: But is this condition really required? I know CUDA applies the same restrictions in `host_config.h`, but that would mean we would have to update CMake everytime CUDA updates their max gcc version.

Comment (pavanky) [inline src/backend/cuda/CMakeLists.txt]: @shehzan10 This is a bit more explicit.

**Comment by umar456** [comment]
Have you guys found any other issues with this PR?

**Comment by pavanky** [comment]
@umar456 just that #1687 should be merged and this should be rebased if there are merge conflicts.

**Comment by 9prady9** [comment]
@arrayfire/core-devel Can this be merged now ? I can use the CUDA backend related changes(C++11 enabled, boost removed) for the thread safety branch i am working on.

**Comment by umar456** [comment]
@pavanky @shehzan10 I think this PR is good to go. Let me know what you think.

**Pull request merged**